### PR TITLE
Add PFTauDiscriminator based on IP variables

### PR DESCRIPTION
**Pull request by sarafiorendi**
Main purpose is to use this producer for the displaced tau triggers in Run3.

### CODE REVIEW

**Comment by azotz** [comment]
Hello @sarafiorendi,
is there supposed to be a config file included with this module?

**Comment by azotz** [comment]
Hello @sarafiorendi,
is there supposed to be a config file included with this module?

**Comment by azotz** [comment]
All normal tests passed. Can you maybe provide a recipe so I can test this module specifically?

**Comment by azotz** [comment]
All normal tests passed. Can you maybe provide a recipe so I can test this module specifically?

File: RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationByIPCut.cc [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+using namespace reco;
+
+class PFRecoTauDiscriminationByIPCut : public PFTauDiscriminationProducerBase {
+public:
+  explicit PFRecoTauDiscriminationByIPCut(const edm::ParameterSet& iConfig)
+      : PFTauDiscriminationProducerBase(iConfig),
+        m_TausTIPToken(consumes<PFTauTIPAssociationByRef>(iConfig.getParameter<edm::InputTag>("TausIP"))),
+        m_tauTIPSelectorString(iConfig.getParameter<std::string>("Cut")),
+        m_tauTIPSelector(m_tauTIPSelectorString) {}
```
Comment (azotz) [inline RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationByIPCut.cc]:
```suggestion
        tausTIPToken_(consumes<PFTauTIPAssociationByRef>(iConfig.getParameter<edm::InputTag>("TausIP"))),
        tauTIPSelectorString_(iConfig.getParameter<std::string>("cut")),
        tauTIPSelector_(m_tauTIPSelectorString) {}
```

Starting data members with a lower case and trailing "_" is the preferred way

File: RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationByIPCut.cc [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  typedef edm::AssociationVector<reco::PFTauRefProd, std::vector<reco::PFTauTransverseImpactParameterRef> >
+      PFTauTIPAssociationByRef;
+  edm::EDGetTokenT<PFTauTIPAssociationByRef> m_TausTIPToken;
+  edm::Handle<PFTauTIPAssociationByRef> TausTIP;
+
+  const std::string m_tauTIPSelectorString;
+  const StringCutObjectSelector<reco::PFTauTransverseImpactParameter> m_tauTIPSelector;
```
Comment (azotz) [inline RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationByIPCut.cc]:
```suggestion
  edm::EDGetTokenT<PFTauTIPAssociationByRef> tausTIPToken_;
  edm::Handle<PFTauTIPAssociationByRef> tausTIP_;

  const std::string tauTIPSelectorString_;
  const StringCutObjectSelector<reco::PFTauTransverseImpactParameter> tauTIPSelector_;
```

File: RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationByIPCut.cc [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+};
+
+double PFRecoTauDiscriminationByIPCut::discriminate(const PFTauRef& thePFTauRef) const {
+  return m_tauTIPSelector(*(*TausTIP)[thePFTauRef]) ? 1. : 0.;
```
Comment (azotz) [inline RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationByIPCut.cc]:
```suggestion
  return m_tauTIPSelector(*(*tausTIP_)[thePFTauRef]) ? 1. : 0.;
```

File: RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationByIPCut.cc [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+void PFRecoTauDiscriminationByIPCut::beginEvent(const edm::Event& event, const edm::EventSetup& eventSetup) {
+  event.getByToken(m_TausTIPToken, TausTIP);
```
Comment (azotz) [inline RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationByIPCut.cc]:
```suggestion
  event.getByToken(tausTIPToken_, tausTIP_);
```

**Comment by sarafiorendi** [comment]
Hi, 
I did not plan to include a cfg file unless is the common policy.
You can test the usage of this module in an HLT path e.g. with

```
cmsrel CMSSW_12_3_0_pre6
cd CMSSW_12_3_0_pre6/src
cmsenv
cp /afs/cern.ch/user/f/fiorendi/public/displacedTaus/forTauPOG/* .
cmsRun customise_hlt_12_3_0.py
```

This will run the double displaced tau HLT path replacing the old dxy filter with the new producer + a selector and filter.

**Comment by azotz** [comment]
> Hi, I did not plan to include a cfg file unless is the common policy. You can test the usage of this module in an HLT path e.g. with
> 
> ```
> cmsrel CMSSW_12_3_0_pre6
> cd CMSSW_12_3_0_pre6/src
> cmsenv
> cp /afs/cern.ch/user/f/fiorendi/public/displacedTaus/forTauPOG/* .
> cmsRun customise_hlt_12_3_0.py
> ```
> 
> This will run the double displaced tau HLT path replacing the old dxy filter with the new producer + a selector and filter.

Thank you for providing a recipe, I will test it.

File: RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationByIPCut.cc [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+void PFRecoTauDiscriminationByIPCut::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>("TausIP", edm::InputTag("hltTauIPCollection"));
+  desc.add<std::string>("Cut", "abs(dxy) > -999.");
```
Comment (azotz) [inline RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationByIPCut.cc]: Is there a need to implement the discrimination requirement as a string? Would it be enough to use `abs(dxy) > X`, where only `X` is a configurable parameter as float/double?

Comment (azotz) [inline RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationByIPCut.cc]:
```suggestion
  desc.add<std::string>("cut", "abs(dxy) > -999.");
```

File: RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationByIPCut.cc [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  {
+    edm::ParameterSetDescription psd0;
+    psd0.add<std::string>("BooleanOperator", "and");
+    desc.add<edm::ParameterSetDescription>("Prediscriminants", psd0);
+  }
+  desc.add<edm::InputTag>("PFTauProducer", edm::InputTag("pfRecoTauProducer"));
+  descriptions.add("pfRecoTauDiscriminationByIPCut", desc);
```
Comment (azotz) [inline RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationByIPCut.cc]:
```suggestion
  descriptions.addWithDefaultLabel(desc);
```

**Comment by azotz** [comment]
> Hi, I did not plan to include a cfg file unless is the common policy. You can test the usage of this module in an HLT path e.g. with
> 
> ```
> cmsrel CMSSW_12_3_0_pre6
> cd CMSSW_12_3_0_pre6/src
> cmsenv
> cp /afs/cern.ch/user/f/fiorendi/public/displacedTaus/forTauPOG/* .
> cmsRun customise_hlt_12_3_0.py
> ```
> 
> This will run the double displaced tau HLT path replacing the old dxy filter with the new producer + a selector and filter.

No need to include a cfg file, this is fine. I would ask you to comment on/implement the proposed changes, so we can proceed with a PR to the official cmssw

**Comment by sarafiorendi** [comment]
hi, I can not see the proposed modifications, which ones are you referring to?

**Comment by azotz** [comment]
> hi, I can not see the proposed modifications, which ones are you referring to?

My bad. They should be visible now

File: RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationByIPCut.cc (hunk 5 above)
Comment (sarafiorendi) [reply RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationByIPCut.cc]: the idea here was that you could also decide to cut on, e.g., abs(dxy)/dxy_err, or other quantities you can access via the PFTauTransverseImpactParameter class. so the string would be more flexible in this sense

File: RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationByIPCut.cc [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+using namespace reco;
+
+class PFRecoTauDiscriminationByIPCut : public PFTauDiscriminationProducerBase {
+public:
+  explicit PFRecoTauDiscriminationByIPCut(const edm::ParameterSet& iConfig)
+      : PFTauDiscriminationProducerBase(iConfig),
+        tausTIPToken_(consumes<PFTauTIPAssociationByRef>(iConfig.getParameter<edm::InputTag>("TausIP"))),
+        tauTIPSelectorString_(iConfig.getParameter<std::string>("cut")),
+        tauTIPSelector_(m_tauTIPSelectorString) {}
```
Comment (azotz) [inline RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationByIPCut.cc]:
```suggestion
        tauTIPSelector_(tauTIPSelectorString_) {}
```

File: RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationByIPCut.cc [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+      PFTauTIPAssociationByRef;
+  edm::EDGetTokenT<PFTauTIPAssociationByRef> tausTIPToken_;
+  edm::Handle<PFTauTIPAssociationByRef> tausTIP_;
+
+  const std::string tauTIPSelectorString_;
+  const StringCutObjectSelector<reco::PFTauTransverseImpactParameter> tauTIPSelector_;
+};
+
+double PFRecoTauDiscriminationByIPCut::discriminate(const PFTauRef& thePFTauRef) const {
+  return m_tauTIPSelector(*(*tausTIP_)[thePFTauRef]) ? 1. : 0.;
```
Comment (azotz) [inline RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationByIPCut.cc]:
```suggestion
  return tauTIPSelector_(*(*tausTIP_)[thePFTauRef]) ? 1. : 0.;
```

File: RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationByIPCut.cc [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+ * class PFRecoTauDiscriminationByIPCut
+ */
+
+using namespace reco;
+
+class PFRecoTauDiscriminationByIPCut : public PFTauDiscriminationProducerBase {
+public:
+  explicit PFRecoTauDiscriminationByIPCut(const edm::ParameterSet& iConfig)
+      : PFTauDiscriminationProducerBase(iConfig),
+        tausTIPToken_(consumes<PFTauTIPAssociationByRef>(iConfig.getParameter<edm::InputTag>("TausIP"))),
```
Comment (azotz) [inline RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationByIPCut.cc]:
```suggestion
        tausTIPToken_(consumes<PFTauTIPAssociationByRef>(iConfig.getParameter<edm::InputTag>("tausTIP"))),
```

File: RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationByIPCut.cc [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+  return tauTIPSelector_(*(*tausTIP_)[thePFTauRef]) ? 1. : 0.;
+}
+
+void PFRecoTauDiscriminationByIPCut::beginEvent(const edm::Event& event, const edm::EventSetup& eventSetup) {
+  event.getByToken(tausTIPToken_, tausTIP_);
+}
+
+void PFRecoTauDiscriminationByIPCut::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>("TausIP", edm::InputTag("hltTauIPCollection"));
```
Comment (azotz) [inline RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationByIPCut.cc]:
```suggestion
  desc.add<edm::InputTag>("tausTIP", edm::InputTag("hltTauIPCollection"));
```

**Comment by azotz** [comment]
This is accepted. I will merge the PR and then squash some of the commits before creating the PR to the central cmssw

**Pull request merged**